### PR TITLE
feat: implement mvi search filters

### PR DESCRIPTION
**Pull request by malandis**
Adds filter expression data types and implementation for `search` and
`searchAndFetchVectors`. Includes search filters for equality checks and logical expressions. A followup PR will include the exhaustive set of comparison operators.

### CODE REVIEW

File: packages/client-sdk-nodejs/src/internal/vector-index-data-client.ts [hunk 1]
```diff
@@ -325,6 +326,85 @@ export class VectorIndexDataClient implements IVectorIndexDataClient {
     }
   }
 
+  private static buildFilterExpression(
```
Comment (malandis) [inline packages/client-sdk-nodejs/src/internal/vector-index-data-client.ts:329]: Would have been nice if this logic could be in one place, but the instantiating the proto classes are completely different between grpc nodejs and web.

File: packages/core/src/messages/vector-index-filters.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    return VectorFilterExpressionFactory.or(this, other);
+  }
+
+  /**
+   * Negates this expression.
+   * @returns The negated expression.
+   */
+  public not(): VectorFilterNotExpression {
+    return VectorFilterExpressionFactory.not(this);
+  }
```
Comment (malandis) [inline packages/core/src/messages/vector-index-filters.ts:31]: These methods allow us to chain expressions with logical operators easier:

```typescript
const e1: VectorFilterExpression;
const e2: VectorFilterExpression;

e1.and(e2);
```

File: packages/core/src/messages/vector-index-filters.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+/**
+ * Factory for creating vector filter expressions.
+ */
+export class VectorFilterExpressionFactory {
```
Comment (malandis) [inline packages/core/src/messages/vector-index-filters.ts:37]: I added this factory class to ease the pain building a nontrivial expression. The hard way (long names and lots of `new`s):

```typescript
new VectorFilterAndExpression(new VectorFilterEqualsExpression('field1', 'value1'), new VectorFilterEqualsExpression('field2', 42))
```

And with these conveniences:
```typescript
import { VectorFilterExpressionFactory as F } from package;

F.equals('field1', 'value1').and(F.equals('field2', 42));
// ...or...
F.and(F.equals('field1', 'value1'), F.equals('field2', 42));
```

**Review by malandis — COMMENTED** [review]
Some comments on design choices for discussion.

File: packages/client-sdk-nodejs/src/internal/vector-index-data-client.ts (hunk 1 above)
Comment (cprice404) [reply packages/client-sdk-nodejs/src/internal/vector-index-data-client.ts:329]: yep.

File: packages/common-integration-tests/src/vector-data-plane.ts [hunk 4]
```diff
@@ -1023,6 +1024,127 @@ export function runVectorDataPlaneTest(
...
+          await sleep(2_000);
+
+          for (const {filterExpression, expectedIds, testCaseName} of [
+            {
+              filterExpression: F.equals('str', 'value1'),
+              expectedIds: ['test_item_1'],
+              testCaseName: 'string equality',
+            },
+            {
+              filterExpression: F.not(F.equals('str', 'value1')),
```
[the remark this answers is not in the document]
Comment (cprice404) [reply packages/common-integration-tests/src/vector-data-plane.ts:1090]: this path seems great to me

File: packages/core/src/messages/vector-index-filters.ts (hunk 3 above)
Comment (cprice404) [reply packages/core/src/messages/vector-index-filters.ts:37]: I'm 💯 on having this factory; I think in modern languages/stdlibs, I've more frequently seen stuff like this exposed by just adding an `s` on the name of the thing, so like we did with `Configurations`, I'd probably name this `VectorFilterExpressions`.

**Review by cprice404 — DISMISSED** [review]
lgtm, left one minor thought on naming that you can take or leave.

lmk if there was any other specific thing in the PR you were hoping for focus on.

File: packages/core/src/messages/vector-index-filters.ts (hunk 3 above)
Comment (malandis) [reply packages/core/src/messages/vector-index-filters.ts]: Done in [c4aeb19](https://github.com/momentohq/client-sdk-javascript/pull/1119/commits/c4aeb19f4b45d8c0343a7f98277c64dc9cfb058c)

**Review by cprice404 — APPROVED** [review]
filter

**Pull request merged**